airbytehq/airbyte
# connectors-ci: test java connectors

## What
Closes #23971
This PR is an attempt at running Java connectors tests with Dagger.
I implemented:
* the logic to run different test suite according to connector language.
* the containerization of Gradle task execution in Dagger

## How to test locally
Follow the instructions to install the CLI [here](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md#L13)
run from Airbyte repo root:
```bash
_EXPERIMENTAL_DAGGER_RUNNER_HOST=docker-image://registry.dagger.io/engine:main@sha256:7983a34a8fbcf6fbed3ec96d462556f67ceff52497cd8b60808359ddb8f867a3 airbyte-ci connectors test --name=source-openweather --name=source-bigquery
```

@dagger folks: you can't run this command if you don't have a `credentials.json` file in `airbyte-integrations/connectors/source-bigquery/secrets`. 
You should create a `credentials.json` file following [this structure](https://gist.github.com/alafanechere/455c68c6253923751206246da65cb798). And run
```bash
_EXPERIMENTAL_DAGGER_RUNNER_HOST=docker-image://registry.dagger.io/engine:main@sha256:7983a34a8fbcf6fbed3ec96d462556f67ceff52497cd8b60808359ddb8f867a3 airbyte-ci connectors --use-remote-secrets=False test --name=source-openweather --name=source-bigquery
```

## 🎉 What is working
* Wrapping Gradle in Dagger and providing Gradle an access to a shared docker host works pretty well. The shared docker host is required to expose docker images built with Gradle to other steps that might require it (Connector Acceptance Test). The Gradle caching works.
* Disabling connector acceptance test run in Gradle in favor of the Dagger AcceptanceTest step works well.
* Running  the`source-postgres` tests using `testcontainers` (which is spawning containers to run test database) is working pretty well. I had to disable the cleanup container (Ryuk) because it is not working in the docker-in-docker context. I'm not sure of the impact there.

## 💔 What is not working
See updates in comments
* ~~The destination `integrationTestJava` Gradle task is failing because the Python Gradle plugin can't find Python in the container. Python is required for java destination build because it depends on normalization build.~~
* ~~I could not run a `integrationTestJava` successfully on Postgres because a couple of tests are not working locally (even with the Gradle command executed locally). I created an issue [here](https://github.com/airbytehq/airbyte/issues/24244) about this problem. @rodireich identified the root cause and will open fix PRs.~~
* ~~A bug in buildkit leads to service container not being cleanly killed, it prevents us from reusing the `dockerd` service container we're relying on to share built images between containers. As a workaround I've downgraded Dagger to `0.4.1` but our infrastructure runs on a more recent version  that depends on a couple of fix / optimization made in `0.4.2`. So we should not merge this [PR](https://github.com/moby/buildkit/pull/3754) until this fix is merged to buildkit and Dagger releases a new version.~~
* ~~The `integrationTestJava` Gradle task can't find the dev connector image and it needs it for a couple of test. I need to investigate more why. I expected this task to run `airbyteDocker` or `build` that will create the dev image on the docker host.~~


## Recommended reading order
### For folks interested in how we run different Step according to connector languages:
https://github.com/airbytehq/airbyte/blob/ee2d3b3f5c5ed1d9b9596130879c76e897f19424/tools/ci_connector_ops/ci_connector_ops/pipelines/tests/__init__.py#L12
https://github.com/airbytehq/airbyte/blob/ee2d3b3f5c5ed1d9b9596130879c76e897f19424/tools/ci_connector_ops/ci_connector_ops/pipelines/tests/java_connectors.py#L99
https://github.com/airbytehq/airbyte/blob/ee2d3b3f5c5ed1d9b9596130879c76e897f19424/tools/ci_connector_ops/ci_connector_ops/pipelines/tests/python_connectors.py#L99
### For folks interested in how we wrap Gradle in Dagger
https://github.com/airbytehq/airbyte/blob/ee2d3b3f5c5ed1d9b9596130879c76e897f19424/tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py#L304
https://github.com/airbytehq/airbyte/blob/ee2d3b3f5c5ed1d9b9596130879c76e897f19424/tools/ci_connector_ops/ci_connector_ops/pipelines/tests/java_connectors.py#L17
